src/libraries/System.Private.CoreLib/src/System/Text/TranscodingStream.cs

```diff
@@ -0,0 +1,598 @@
+
+        protected override void Dispose(bool disposing)
+        {
+            Debug.Assert(disposing, "This type isn't finalizable.");
+
+            if (_innerStream is null)
+            {
+                return; // dispose called multiple times, ignore
+            }
+
+            // First, flush any pending data to the inner stream.
+
+            ArraySegment<byte> pendingData = FinalFlushWriteBuffers();
+            if (pendingData.Count != 0)
+            {
+                _innerStream.Write(pendingData);
+            }
+
+            // Mark our object as disposed
+
+            Stream innerStream = _innerStream;
+            _innerStream = null;
+
+            // And dispose the inner stream if needed
+
+            if (!_leaveOpen)
+            {
+                innerStream.Dispose();
+            }
+        }
+
+        public override ValueTask DisposeAsync()
+        {
+            if (_innerStream is null)
+            {
+                return default; // dispose called multiple times, ignore
+            }
+
+            // First, get any pending data destined for the inner stream.
+
+            ArraySegment<byte> pendingData = FinalFlushWriteBuffers();
+
+            if (pendingData.Count == 0)
+            {
+                // Fast path: just dispose of the object graph.
+                // No need to write anything to the stream first.
+
+                Stream innerStream = _innerStream;
+                _innerStream = null;
+
+                return (_leaveOpen)
+                    ? default /* no work to do */
+                    : innerStream.DisposeAsync();
+            }
+
+            // Slower path; need to perform an async write followed by an async dispose.
+
+            return DisposeAsyncCore(pendingData);
+            async ValueTask DisposeAsyncCore(ArraySegment<byte> pendingData)
+            {
+                Debug.Assert(pendingData.Count != 0);
+
+                Stream innerStream = _innerStream;
+
+                await innerStream.WriteAsync(pendingData.AsMemory()).ConfigureAwait(false);
+                _innerStream = null;
+
+                if (!_leaveOpen)
+                {
+                    await innerStream.DisposeAsync().ConfigureAwait(false);
+                }
+            }
+        }
+
+        public override int EndRead(IAsyncResult asyncResult)
+            => TaskToApm.End<int>(asyncResult);
+
+        public override void EndWrite(IAsyncResult asyncResult)
+            => TaskToApm.End(asyncResult);
+
+#pragma warning disable CS3016 // Arrays as attribute arguments is not CLS-compliant
+#pragma warning disable CS8774 // Member must have a non-null value when exiting.
+
+        // Sets up the data structures that are necessary before any read operation takes place,
+        // throwing if the object is in a state where reads are not possible.
+        [MemberNotNull(nameof(_innerStream), nameof(_innerDecoder), nameof(_thisEncoder))]
```

Comment:
FWIW, once we get an updated compiler, you'll be able to write this as multiple MemberNotNulls, one for each field, and you won't need the suppression for CS3016.